xtra-libp2p/src/endpoint.rs

```diff
@@ -250,10 +259,37 @@ impl Endpoint {
             inflight_connections: HashSet::default(),
             connection_timeout,
             subscribers,
+            peer_listen_protocols: HashMap::default(),
+        }
+    }
+
+    fn does_peer_listen_for(&self, peer_id: PeerId, protocols: &[&str]) -> Result<(), Error> {
+        let listen_protocols = match self.peer_listen_protocols.get(&peer_id) {
+            Some(listen_protocols) => listen_protocols,
+            None => {
+                // Without the information about the peer we cannot decide either way. We default to
+                // answering yes to our caller, just in case
+                return Ok(());
+            }
+        };
+        let listen_protocols = listen_protocols
+            .iter()
+            .map(|p| p.as_ref())
+            .collect::<Vec<_>>();
+
+        if !protocols
+            .iter()
+            .any(|protocol| listen_protocols.contains(protocol))
+        {
+            return Err(Error::ProtocolNotSupportedByPeer);
         }
+
+        Ok(())
```

Comment:
🙃 imho a bit more readable version (and a one-liner ;D) 

```suggestion
        protocols
            .iter()
            .any(|protocol| listen_protocols.contains(protocol))
            .then(|| ())
            .ok_or(Error::ProtocolNotSupportedByPeer)
```

Reply:
I had this in mind at first but I decided against it because `.then(|| ())` is not so obvious. But given that you like it I will write it like this!